core/finality-grandpa/src/observer.rs

```diff
@@ -210,85 +272,105 @@ pub fn run_grandpa_observer<B, E, Block: BlockT<Hash=H256>, N, RA, SC>(
-
-					set_state
-				},
-				VoterCommand::ChangeAuthorities(new) => {
-					// start the new authority set using the block where the
-					// set changed (not where the signal happened!) as the base.
-					let set_state = VoterSetState::live(
-						new.set_id,
-						&*authority_set.inner().read(),
-						(new.canon_hash, new.canon_number),
-					);
-
-					#[allow(deprecated)]
-					crate::aux_schema::write_voter_set_state(&**client.backend(), &set_state)?;
-
-					set_state
-				},
-			};
+		self.observer = Box::new(observer);
+	}
 
-			Ok(FutureLoop::Continue((authority_set, consensus_changes, set_state.into(), voter_commands_rx)))
-		};
+	fn handle_voter_command(
+		&mut self,
+		command: VoterCommand<B::Hash, NumberFor<B>>,
+	) -> Result<(), Error> {
+		// the observer doesn't use the voter set state, but we need to
+		// update it on-disk in case we restart as validator in the future.
+		self.persistent_data.set_state = match command {
+			VoterCommand::Pause(reason) => {
+				info!(target: "afg", "Pausing old validator set: {}", reason);
 
-		// run observer and listen to commands (switch authorities or pause)
-		observer.select2(voter_commands_rx).then(move |res| match res {
-			Ok(future::Either::A((_, _))) => {
-				// observer commit stream doesn't conclude naturally; this could reasonably be an error.
-				Ok(FutureLoop::Break(()))
-			},
-			Err(future::Either::B(_)) => {
-				// the `voter_commands_rx` stream should not fail.
-				Ok(FutureLoop::Break(()))
+				let completed_rounds = self.persistent_data.set_state.read().completed_rounds();
+				let set_state = VoterSetState::Paused { completed_rounds };
+
+				#[allow(deprecated)]
+				crate::aux_schema::write_voter_set_state(&**self.client.backend(), &set_state)?;
+
+				set_state
 			},
-			Ok(future::Either::B(((None, _), _))) => {
-				// the `voter_commands_rx` stream should never conclude since it's never closed.
-				Ok(FutureLoop::Break(()))
+			VoterCommand::ChangeAuthorities(new) => {
+				// start the new authority set using the block where the
+				// set changed (not where the signal happened!) as the base.
+				let set_state = VoterSetState::live(
+					new.set_id,
+					&*self.persistent_data.authority_set.inner().read(),
+					(new.canon_hash, new.canon_number),
+				);
+
+				#[allow(deprecated)]
+				crate::aux_schema::write_voter_set_state(&**self.client.backend(), &set_state)?;
+
+				set_state
 			},
-			Err(future::Either::A((CommandOrError::Error(e), _))) => {
+		}.into();
+
+		self.rebuild_observer();
+		Ok(())
+	}
+}
+
+impl<B, N, E, Bk, RA> Future for ObserverWork<B, N, E, Bk, RA>
+where
+	B: BlockT<Hash=H256>,
+	N: Network<B>,
+	N::In: Send + 'static,
+	NumberFor<B>: BlockNumberOps,
+	RA: 'static + Send + Sync,
+	E: CallExecutor<B, Blake2Hasher> + Send + Sync + 'static,
+	Bk: Backend<B, Blake2Hasher> + 'static,
+{
+	type Item = ();
+	type Error = Error;
+
+	fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+		match self.observer.poll() {
```

Comment:
Why is `observer` a field on `ObserverWork` rather than constructed on demand at the beginning of `poll()`?

Reply:
I didn't look at the code of the `observer`, but the previous code behaves in the same way.

Reply:
Maybe I don't understand. The old code looks like it creates a new `grandpa_observer` on each execution of the closure. And in the current code, `rebuild_observer` creates the observer from scratch and drops the old one, so it seems like all of the state is given in the args to `grandpa_observer`.

Reply:
It is actually hard to understand because the code relies on implicit side-effects. We only need to `rebuild_observer` when we handle a new command, because the data in https://github.com/paritytech/substrate/pull/3309/files#diff-2fb237d4f37ba90960eacc42fbc35f6aR244 will have changed (from either importing or finalizing a block in another task). Maybe we can add a comment there to explain what is going on here.

Reply:
> The old code looks like it creates a new grandpa_observer on each execution of the closure.

The closure is called at each iteration of the loop, which happens after either the observer or `voter_commands_rx` produce a value.

The old code is indeed extremely confusing, and that's partially why I think it's a good idea to rewrite it.

Reply:
> Maybe we can add a comment to explain what is going on here.

What do you think I should comment on?
I'm not going to explain that we don't recreate the observer every time we poll, because the obvious way of doing it is indeed to not recreate the observer every time we poll.